src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,16 +33,28 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        internal static T EnsureInitializedDisposable<T>([AllowNull] ref T? field, Func<T> initializer) where T : class, IDisposable
+            => Volatile.Read(ref field) ?? EnsureInitializedDisposableCore(ref field, initializer);
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+        private static T EnsureInitializedDisposableCore<T>([AllowNull] ref T? field, Func<T> initializer) where T : class, IDisposable
+        {
+            T value = initializer();
+
+            if (Interlocked.CompareExchange(ref field, value, null) != null)
+            {
+                value.Dispose();
+            }
+
+            return field;
+        }
+
+        public static TextReader In => EnsureInitializedDisposable(ref s_in, () => ConsolePal.GetOrCreateReader());
 
         public static Encoding InputEncoding
         {
             get
             {
-                return EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
+                return LazyInitializer.EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
```

Comment:
The `LazyInitializer` looks a bit nicer, but it is many times more expensive compared to just writing the Interlocked.CompareExchange directly. I would just use `Interlocked.CompareExchange` directly here. Console is used by pretty much every program

Reply:
I've localized the this to `Console`. I don't understand why using `LazyInitializer` is more costly. Is it because these things get compiled r2r?

Reply:
LazyInitalizer requires delegate, method backing the delegate, static to cache the delegate, and several generic method instantiations. It does not have much to do with R2R.

Reply:
ok, I have some work still to get rid of the delegate. I was making wrong assumptions on where the cost was.